contracts/token/BiconomyTokenPaymaster.sol

```diff
@@ -392,20 +344,29 @@ contract BiconomyTokenPaymaster is BasePaymaster, ReentrancyGuard, TokenPaymaste
         uint256 actualGasCost
     ) internal virtual override {
 
-        (address account, IERC20 feeToken, ExchangeRateSource priceSource, uint256 exchangeRate, uint256 fee, uint256 gasPriceUserOp) = abi
-            .decode(context, (address, IERC20, ExchangeRateSource, uint256, uint256, uint256));
+        (address account, IERC20 feeToken, ExchangeRateSource priceSource, uint256 exchangeRate, uint256 fee, uint256 maxFeePerGas, uint256 maxPriorityFeePerGas) = abi
+            .decode(context, (address, IERC20, ExchangeRateSource, uint256, uint256, uint256, uint256));
 
         uint256 effectiveExchangeRate = exchangeRate;
 
         if (priceSource == ExchangeRateSource.CHAINLINK_PRICE_ORACLE_BASED) {
-            effectiveExchangeRate = this.exchangePrice(address(feeToken));
+            effectiveExchangeRate = exchangePrice(address(feeToken));
         } 
 
+        uint256 gasPriceUserOp = maxFeePerGas;
+
+        // review Could do below if we're okay to touch BASEFEE in postOp call
+        unchecked {
+            if (maxFeePerGas != maxPriorityFeePerGas) {
+                gasPriceUserOp = Math.min(maxFeePerGas, maxPriorityFeePerGas + block.basefee);
+            }
+        }
+
         uint256 actualTokenCost = ((actualGasCost + (UNACCOUNTED_COST * gasPriceUserOp)) * effectiveExchangeRate) / 1e18;
         if (mode != PostOpMode.postOpReverted) {
-            // review if below silently fails should notify in event accordingly
+            // review if below fails should notify in event / revert at the risk of reputation
             feeToken.safeTransferFrom(account, feeReceiver, actualTokenCost + fee);
-            emit TokenPaymasterOperation(account, address(feeToken), actualTokenCost + fee);
+            emit TokenPaymasterOperation(account, address(feeToken), actualTokenCost, fee);
```

Comment:
Also emit `effectiveExchangePrice` , `userOpHash`, `priceSource` in the event for better tracking purpose.


If possible add another field which have value 'actualTokenCost + fee' also to simplify the accounting if it is to be used off-chain.

Need to pass userOpHash in the context from `validatePaymasterUserOp` method.

Reply:
so what all we should emit?

userOpHash will anyways be emitted in UserOperationEvent (sender as well)

I am also emitting actualTokenCost and fee separately, do you want us to emit actualTokenCost, fee, and actualTokenCost + fee ? 

and should all of these be in same event (I can index at max 3 values - all indexed is ideal)

Reply:
Yes we need to also emit UserOpHash in this event, coz we want to co-relate the transaction when we update these transactions in the database in paymaster service.
Eg. When signing request comes to paymaster service, we save the data in db with userOpHash as primary key.
Now when we track the TokenPaymasterOperation event, then it becomes hard to corelate it with userOpHash already present in the database. Now we have to get all the logs of same transactino and try to search UserOperation event (which can be multiple btw in same transaction) and then try to corelate it with this event.


That's why its important to emit userOpHash also in this event, so we can corelate this event with actual transaction data in database.

Also its fine if there are more than  3 events, it can be decoded using the ABI, not a big issue.

So in all we need to emit
userOpHash, (actualTokenCost+fee), fee, account, tokenAddress, effectiveExchangePrice, priceSource

Where account, tokenAddress, actualTokenCost+fee are indexed.

Reply:
ok, let me make single event and add more fields.

Reply:
done